cli_tools/common/utils/daisyutils/daisy_worker.go

```diff
@@ -74,35 +78,93 @@ func createResourceLabelerIfMissing(env EnvironmentSettings, hooks []interface{}
 }
 
 type defaultDaisyWorker struct {
-	wf     *daisy.Workflow
-	logger logging.Logger
-	env    EnvironmentSettings
-	hooks  []interface{}
+	workflowProvider WorkflowProvider
+	finished         *daisy.Workflow
+	logger           logging.Logger
+	env              EnvironmentSettings
+	hooks            []interface{}
+
+	cancel      chan string
+	cancelGuard sync.Once
 }
 
 // Run runs the daisy workflow with the supplied vars.
-func (w *defaultDaisyWorker) Run(vars map[string]string) error {
-	if err := (&ApplyAndValidateVars{w.env, vars}).PreRunHook(w.wf); err != nil {
-		return err
+func (w *defaultDaisyWorker) Run(vars map[string]string) (err error) {
+	var wf *daisy.Workflow
+	for attempts := 0; attempts <= 1; attempts++ {
+		if wf, err = w.workflowProvider(); err != nil {
+			break
+		}
+		if err = w.checkIfCancelled(wf); err != nil {
+			break
+		}
+		var retryRequested bool
+		retryRequested, err = w.runOnce(wf, vars)
+		if err == nil || !retryRequested {
+			break
+		}
+		w.logger.Debug("Retry requested. err=" + err.Error())
```

Comment:
If `err` is not `nil`, but `retryRequested` is `false`, the execution will still end up here, so printing _Retry requested_ is not correct.

Reply:
Ahh, good point. Updated the log message.